cfpb/cfgov-refresh
# Conference registration form improvements

This PR makes the following improvements to the conference registration form:

1. Customizable success message (`RichTextBlock`), was previously hardcoded.
2. Use a query string parameter on the page URL (`?success`) to trigger the success instead of using Django messages, which requires an additional cookie.
3. Customizable failure message (`CharBlock`) when something (GovDelivery signup) goes wrong.

## Additions

- New customizable form blocks for success message and failure message.

## Removals

- Remove use of the [Django messages framework](https://docs.djangoproject.com/en/1.8/ref/contrib/messages/) to store and display success or error messages.

## Changes

- Use a Django form non-field error to display the failure message when something goes wrong with GovDelivery, instead of using Django messages.
- Use a query string parameter to display the success message when registration succeeds, isntead of using Django messages.

## Testing

- Run a local server, making sure you have valid GovDelivery settings in your environment.
- Create a new `BrowsePage` in the Wagtail admin. Give it a title and add a `ConferenceRegistrationForm` block.
- Enter a heading, an INVALID GovDelivery code, a few session names, a capacity of 0, a success message (including HTML content), an at capacity message, and a failure message.
- Preview the page. You should see the at capacity message, because the capacity was set to 0.
- Next, increase the capacity to a positive integer and publish, and view the page.
- Try submitting the form with invalid entries, say `foo@bar` for the email, or without selecting any sessions. You'll see the form come back with some errors.
- Try submitting the form with valid entries. You'll see the form come back with the failure message, because of your invalid GovDelivery code.
- Go back and edit your form to input a valid GovDelivery code. Reload the page and resubmit the form with valid entries. You'll see the form come back with the success message. Optionally check your local database (`data_research.ConferenceRegistration`) to verify that your information was recorded properly.

## Review

- @cfarm @marteki @nataliafitzgerald @cfpb/cfgov-backends 

## Screenshots

Form field errors:
![image](https://cloud.githubusercontent.com/assets/654645/20460257/9eb99d54-aeab-11e6-950b-3dd35998f2da.png)

GovDelivery failure message (configurable):
![image](https://cloud.githubusercontent.com/assets/654645/20460263/c0de8da4-aeab-11e6-8338-131989c591e1.png)

Success message (configurable):
![image](https://cloud.githubusercontent.com/assets/654645/20460276/2c050554-aeac-11e6-83f6-a1f054e26279.png)

Success message source (@cfarm you mentioned a potential tweak to make this line up better?):
```html
<div class="m-notification m-notification__success m-notification__visible">
    <span class="m-notification_icon cf-icon"></span>
    <div class="m-notification_content">
        <p class="h4 m-notification_message"></p>
        <div class="rich-text">
            <p>Success! Thanks for registering.</p>
            <p>You should hear from us soon, but if not, you can email us at <a href="mailto:hello@cfpb.gov">hello@cfpb.gov</a>.</p>
        </div>
        <p></p> 
    </div>
</div>
```

## Notes

- Using a query string parameter to indicate success means that if a registration form is on a page like `/conference/register/`, then on successful form submission, the user is redirected to `/confererence/register/?success`. This avoids having to use the Django messages framework, which can complicate site caching (due to use of cookies and restrictions on HTML content in the cookied message content). It has one possible downside, which is that the success page (`/conference/register/?success`) is publicly available to anyone who should stumble on the URL. @rosskarchner any concerns about this alternate approach?

## Checklist
